dhis-2/dhis-support/dhis-support-cache-invalidation/src/main/java/org/hisp/dhis/cacheinvalidation/DebeziumService.java

```diff
@@ -0,0 +1,229 @@
+        // "dhis2_[EPOCH_SECONDS]_[OPTIONAL_CONFIG_ID]_[UUID4]
+        long timeNow = System.currentTimeMillis() / 1000L;
+        String slotName = "dhis2_" + timeNow + "_" + configSlotName + "_"
+            + UUID.randomUUID().toString().replaceAll( "-", "" );
+
+        Properties props = Configuration.create().build().asProperties();
+        props.setProperty( "name", slotName );
+        props.setProperty( "slot.name", slotName );
+        props.setProperty( "plugin.name", "pgoutput" );
+        props.setProperty( "publication.autocreate.mode", "filtered" );
+        props.setProperty( "connector.class", "io.debezium.connector.postgresql.PostgresConnector" );
+        props.setProperty( "offset.storage", "org.apache.kafka.connect.storage.MemoryOffsetBackingStore" );
+        props.setProperty( "offset.flush.interval.ms", "60000" );
+        props.setProperty( "database.tcpKeepAlive", "true" );
+        props.setProperty( "database.hostname", dbHostname );
+        props.setProperty( "database.port", dbPort );
+        props.setProperty( "database.user", username );
+        props.setProperty( "database.password", password );
+        props.setProperty( "database.server.name", dbName );
+        props.setProperty( "database.dbname", dbName );
+        props.setProperty( "snapshot.mode", "never" );
+        props.setProperty( "slot.drop.on.stop", "true" );
+        props.setProperty( "table.exclude.list", "public.spatial_ref_sys,public.audit," + excludeList );
+
+        engine = DebeziumEngine
+            .create( ChangeEventFormat.of( Connect.class ) )
+            .using( props )
+            .using( this::handleCompletionCallback )
+            .using( connectorCallback )
+            .notifying( dbChangeEventHandler::handleDbChange )
+            .build();
+
+        executor.execute( engine );
+
+        if ( !engineStartupLatch.await( STARTUP_WAIT_TIMEOUT_SECONDS, TimeUnit.SECONDS ) )
+        {
+            log.error( "Debezium engine startup timeout exceeded! " + SHUTDOWN_DISABLED_MSG );
+
+            if ( shutdownOnConnectorStop )
+            {
+                shutdown( "Debezium engine startup failed! " + SEE_LOG_MSG, SHUTDOWN_DELAY_SECONDS, null );
+            }
+        }
+        else
+        {
+            log.info( "Debezium engine started successfully!" );
+        }
+    }
+
+    /**
+     * Important that this is called before shutdown so Debezium can remove its
+     * replication slot.
+     */
+    @PreDestroy
+    public void stopDebezium()
+        throws IOException
+    {
+        log.info( "Stopping the Debezium engine..." );
+        engine.close();
+    }
+
+    private void shutdown( String msg, Long countDownSeconds, Throwable throwable )
+    {
+        if ( countDownSeconds != null )
+        {
+            final ScheduledExecutorService scheduler = Executors.newScheduledThreadPool( 1 );
+            scheduler.schedule( () -> shutdown( msg, null, throwable ), countDownSeconds, TimeUnit.SECONDS );
```

Comment:
Would be more clear to have two methods, one that is scheduled and one doing the work called by the scheduled one.